### PR TITLE
BUG: lib.infer_dtype with mixed-freq Periods

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Also adds a convert_period option to lib.maybe_infer_objects

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2343,14 +2367,6 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=False,
         elif util.is_float_object(val):
             floats[i] = complexes[i] = val
             seen.float_ = True
-        elif util.is_datetime64_object(val):
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:2346]: unrelated to the rest of the PR, this chunk of code is just redundant with a chunk below

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1996,18 +1998,35 @@ cpdef bint is_time_array(ndarray values, bint skipna=False):
     return validator.validate(values)
 
 
-cdef class PeriodValidator(TemporalValidator):
-    cdef inline bint is_value_typed(self, object value) except -1:
-        return is_period_object(value)
+cdef bint is_period_array(ndarray[object] values):
```
Comment (jreback) [inline pandas/_libs/lib.pyx:2001]: can you not keep the current PeriodValidator format itself? (e.g. just put this function in validate). this is breaking the pattern.

**Review by jreback — CHANGES_REQUESTED** [review]
does this have any user facing changes?

**Comment by jbrockmendel** [comment]
> does this have any user facing changes?

`replace` handles periods better

**Comment by jreback** [comment]
> > does this have any user facing changes?
> 
> `replace` handles periods better

kk can you add a note.

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:2001]: this is the same pattern we use for is_datetime_with_singletz_array (and will end up using for is_interval_array in an upcoming PR)

**Comment by jbrockmendel** [comment]
whatsnew added + green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**